src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -2023,15 +2029,35 @@ public Interop.Sys.SocketEvents HandleSyncEventsSpeculatively(Interop.Sys.Socket
             return events;
         }
 
-        public unsafe void HandleEvents(Interop.Sys.SocketEvents events)
+        // Called on the epoll thread.
+        public Interop.Sys.SocketEvents HandleSyncAndInlineEvents(Interop.Sys.SocketEvents events)
         {
             if ((events & Interop.Sys.SocketEvents.Error) != 0)
             {
-                // Set the Read and Write flags as well; the processing for these events
+                // Set the Read and Write flags; the processing for these events
                 // will pick up the error.
+                events ^= Interop.Sys.SocketEvents.Error;
                 events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
             }
 
+            if ((events & Interop.Sys.SocketEvents.Read) != 0 &&
+                _receiveQueue.ProcessSyncEventOrGetAsyncEvent(this, inlineAsyncEvents: true) == null)
+            {
+                events ^= Interop.Sys.SocketEvents.Read;
+            }
+
+            if ((events & Interop.Sys.SocketEvents.Write) != 0 &&
+                _sendQueue.ProcessSyncEventOrGetAsyncEvent(this, inlineAsyncEvents: true) == null)
+            {
+                events ^= Interop.Sys.SocketEvents.Write;
+            }
+
+            return events;
+        }
+
+        // Called on ThreadPool thread.
+        public unsafe void HandleEvents(Interop.Sys.SocketEvents events)
+        {
```

Comment:
The diff makes it hard to see, but I've removed the mapping from Error to Read/Write here since that has already happened when `HandleEvents` is called on ThreadPool, by earlier calls to `HandleSyncEventsSpeculatively`/`HandleSyncAndInlineEvents`.

Reply:
Maybe assert that the `Error` bit is not set